### PR TITLE
Add Cosmos back-sync script for rollback (PR-D of 4)

### PR DESCRIPTION
**Pull request by geevensingh**
## Summary

PR-D of 4 (region migration east US 2 -> west US 2 + rename dev -> prod).

Adds `scripts/cosmos-back-sync.mjs` - a Node.js script that back-syncs writes from the NEW Cosmos account to the OLD account during the 7-day post-cutover rollback window. **The script does NOT execute on merge** - it's insurance for the rollback playbook. Merging is zero-impact.

The per-doc strategy GETs each new-account doc on the OLD account first, then decides: CREATE (if 404), UPDATE with `If-Match` on the OLD `_etag` (if old is older), or skip + log (if old is fresher). A naive `If-Match` on the NEW account's `_etag` would 412 on every doc; this strategy handles the case where the OLD account had writes during the rollback window.

Conflicts (412 from concurrent writes, or old-fresher cases) are logged to `cosmos-back-sync-conflicts-<ISO8601>.jsonl` in the operator's CWD for manual reconciliation.

## Changes

- `scripts/cosmos-back-sync.mjs` (new): rollback back-sync script.
- `scripts/cosmos-back-sync.test.mjs` (new): unit tests for arg parsing, container set, conflict-file naming, per-doc decision logic.
- `package.json`: add `@azure/cosmos@^4.9.3` to `devDependencies` (matches api workspace pin). Dev-only - doesn't ship to production.
- `package-lock.json`: regenerated for the new dep.

## Validation

- `npm run lint:all`: pass (includes `lint:lockfile`)
- `npm run test:scripts`: pass (new test included)
- `npm test`: pass
- `npm run build`: pass

## SemVer

No bump (script-only; no runtime / user-visible changes).

---
**Session**
- AI-Local: d1db4668-c1d4-43a4-9c76-e86ec82bbb07
- AI-Cloud: fa8e1735-48f9-47cc-9757-4eca1b12f784

### CODE REVIEW

File: scripts/cosmos-back-sync.mjs [hunk 1]
```diff
@@ -0,0 +1,691 @@
...
+    docsProcessed: 0,
+    docsCreated: 0,
+    docsUpdated: 0,
+    docsSkippedOldFresher: 0,
+    conflicts: 0,
+  };
+
+  const iterator = sourceContainer.items.getChangeFeedIterator({
+    changeFeedStartFrom: ChangeFeedStartFrom.Time(new Date(cutoverInstantUnixSeconds * 1000)),
+  });
```
[the remark this answers is not in the document]
Comment (geevensingh) [reply scripts/cosmos-back-sync.mjs:783]: Addressed in 2a73888 (merge: 37513de).

The change-feed-misses-deletes failure mode is real and unfixable
from the back-sync side alone (retroactive AllVersionsAndDeletes
requires pre-cutover `changeFeedPolicy.retentionDuration > 0` on
each container, which was not done before the migration). Adopting
the bot's preferred fix would have required a separate
infrastructure change that doesn't help the existing one-shot
rollback artifact.

What we did instead:

1. The script is no longer silent about this. It now requires an
   explicit `--accept-delete-loss` opt-in flag; without it, the
   script exits 1 before opening any Cosmos session.
2. The header docstring and `USAGE` block call out the limitation
   in a "Known limitation - deletes are not replayed" section.
3. `docs/migration-westus2.md` "Rollback data-loss budget" now
   includes a 4th bullet enumerating the affected containers
   (`blobs`, `history`, `rule-sets`) and recommends post-back-sync
   diff reconciliation against the pre-cutover snapshot before
   declaring the rollback complete.
4. `warnDeletesNotReplayed(stderr)` extracted as a directly-
   testable helper; emit shape is asserted in
   `scripts/cosmos-back-sync.test.mjs`.

The three-agent rubber-duck panel (skeptic + advocate + architect)
confirmed this disposition: the bot's preferred fix is mechanically
infeasible for this rollback window without pre-cutover container
reprovisioning, so the best we can do is make the limitation loud
and require operator acknowledgment.

A follow-up issue will be filed to enable `AllVersionsAndDeletes`
retention on the production containers in advance of any future
migration so this gap closes for the next attempt.

[the remark this answers is not in the document]
Comment (geevensingh) [reply scripts/cosmos-back-sync.mjs:783]: Addressed in 2a73888 (merge: 37513de).

The change-feed-misses-deletes failure mode is real and unfixable
from the back-sync side alone (retroactive AllVersionsAndDeletes
requires pre-cutover `changeFeedPolicy.retentionDuration > 0` on
each container, which was not done before the migration). Adopting
the bot's preferred fix would have required a separate
infrastructure change that doesn't help the existing one-shot
rollback artifact.

What we did instead:

1. The script is no longer silent about this. It now requires an
   explicit `--accept-delete-loss` opt-in flag; without it, the
   script exits 1 before opening any Cosmos session.
2. The header docstring and `USAGE` block call out the limitation
   in a "Known limitation - deletes are not replayed" section.
3. `docs/migration-westus2.md` "Rollback data-loss budget" now
   includes a 4th bullet enumerating the affected containers
   (`blobs`, `history`, `rule-sets`) and recommends post-back-sync
   diff reconciliation against the pre-cutover snapshot before
   declaring the rollback complete.
4. `warnDeletesNotReplayed(stderr)` extracted as a directly-
   testable helper; emit shape is asserted in
   `scripts/cosmos-back-sync.test.mjs`.

The three-agent rubber-duck panel (skeptic + advocate + architect)
confirmed this disposition: the bot's preferred fix is mechanically
infeasible for this rollback window without pre-cutover container
reprovisioning, so the best we can do is make the limitation loud
and require operator acknowledgment.

A follow-up issue will be filed to enable `AllVersionsAndDeletes`
retention on the production containers in advance of any future
migration so this gap closes for the next attempt.

File: scripts/cosmos-back-sync.mjs [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+
+  if (missingArguments.length > 0) {
+    throw new UsageError(`Missing required arguments: ${missingArguments.join(', ')}`);
+  }
+
+  const cutoverInstantUnixSeconds = Number.parseInt(cutoverInstantRaw, 10);
+  if (!Number.isSafeInteger(cutoverInstantUnixSeconds) || cutoverInstantUnixSeconds < 0) {
+    throw new UsageError(
+      'The --cutover-instant-unix-seconds value must be a non-negative integer.',
+    );
```
[the remark this answers is not in the document]
Comment (geevensingh) [reply scripts/cosmos-back-sync.mjs:207]: Addressed in 2a73888 (merge: 37513de).

`parseCliOptions` now validates the raw `--cutover-instant-unix-seconds`
string against `/^\d+$/u` before any `Number()` coercion. This
mirrors how the api workspace's hand-rolled `assert<Shape>` guards
in `api/src/shared/preferences.ts` validate primitive inputs.

Edge cases now covered by tests:

- `'1717200000abc'`  -> reject (trailing garbage, original bot example)
- `'1717200000.9'`   -> reject (decimal)
- `'1e9'`            -> reject (exponential form; `parseInt` would have returned 1)
- `' 1717200000 '`   -> reject (whitespace)
- `'NaN'`            -> reject
- `'Infinity'`       -> reject
- `'+1717200000'`    -> reject (explicit positive sign)
- `'1717200000'`     -> accept
- `'01717200000'`    -> accept (leading zeros legitimate for unix
                       timestamps that happen to start with `0`)

The regex test case list and the digit-only acceptance test in
`scripts/cosmos-back-sync.test.mjs` enumerate all 9 cases above.

[the remark this answers is not in the document]
Comment (geevensingh) [reply scripts/cosmos-back-sync.mjs:207]: Addressed in 2a73888 (merge: 37513de).

`parseCliOptions` now validates the raw `--cutover-instant-unix-seconds`
string against `/^\d+$/u` before any `Number()` coercion. This
mirrors how the api workspace's hand-rolled `assert<Shape>` guards
in `api/src/shared/preferences.ts` validate primitive inputs.

Edge cases now covered by tests:

- `'1717200000abc'`  -> reject (trailing garbage, original bot example)
- `'1717200000.9'`   -> reject (decimal)
- `'1e9'`            -> reject (exponential form; `parseInt` would have returned 1)
- `' 1717200000 '`   -> reject (whitespace)
- `'NaN'`            -> reject
- `'Infinity'`       -> reject
- `'+1717200000'`    -> reject (explicit positive sign)
- `'1717200000'`     -> accept
- `'01717200000'`    -> accept (leading zeros legitimate for unix
                       timestamps that happen to start with `0`)

The regex test case list and the digit-only acceptance test in
`scripts/cosmos-back-sync.test.mjs` enumerate all 9 cases above.

File: scripts/cosmos-back-sync.mjs [hunk 3]
```diff
@@ -0,0 +1,691 @@
...
+        oldDoc: destinationDocument,
+        newDoc: sourceDocument,
+      }),
+    );
+  } catch (error) {
+    const errorCode = getErrorCode(error);
+    const reason =
+      errorCode === PRECONDITION_FAILED_STATUS_CODE || errorCode === CONFLICT_STATUS_CODE
+        ? 'concurrent-write'
+        : 'unknown';
```
[the remark this answers is not in the document]
Comment (geevensingh) [reply scripts/cosmos-back-sync.mjs:711]: Addressed in 2a73888 (merge: 37513de).

Extended `getErrorCode` at the helper level (not the call site) via
a frozen alias map. The skeptic panelist flagged that fixing only
the line-525 site missed a parallel silent-data-loss bug at
line 318 (`readDestinationDocument`): a string-shape `'NotFound'`
would have skipped the document instead of treating it as missing,
silently dropping it from the back-sync.

Adopted normalization:

- `'PreconditionFailed'` -> 412 (matches existing api precedent at
  `api/src/shared/cosmos.ts:159`)
- `'NotFound'` -> 404 (closes the line-318 silent-data-loss vector)

Deferred:

- `'Conflict'` -> 409 -- per the skeptic's evidence standard, no
  observed SDK emission of this exact string shape today. Easy to
  add later if it surfaces; the alias map is one-line extension.

Implementation notes:

- The alias map uses `Object.freeze({ __proto__: null, ... })` so
  an attacker-controlled `code: '__proto__'` follows neither
  `Object.prototype`'s setter nor a real entry. The `in` operator +
  bracket access remain safe.
- Bidirectional `SYNC WITH` JSDoc cross-reference added between
  `scripts/cosmos-back-sync.mjs:getErrorCode` and
  `api/src/shared/cosmos.ts:isCosmosPreconditionFailed`. Future SDK
  changes to string-code shape become mechanically discoverable via
  grep.
- Lumping (412 || 409) happens at the call site, not the helper, so
  the helper stays single-responsibility (numeric coercion) and
  callers retain control over which codes constitute a "conflict"
  for their context.

A follow-up issue will be filed to consider lifting
`isCosmosPreconditionFailed` to a shared cross-workspace package
once a second cross-workspace consumer appears.

File: scripts/cosmos-back-sync.mjs [hunk 4]
```diff
@@ -0,0 +1,883 @@
...
+      );
+      return;
+    }
+
+    await destinationContainer.item(id, partitionKey).replace(sourceDocument, {
+      accessCondition: {
+        type: IF_MATCH,
+        condition: destinationDocument._etag,
+      },
+    });
```
[the remark this answers is not in the document]
Comment (geevensingh) [reply scripts/cosmos-back-sync.mjs:691]: Good catch — the write path was indeed uncovered. Round-2 panel rubber-duck (skeptic + advocate + architect) expanded the original ask into a 13-test container-stubbed suite that pins both the call-shape regressions you flagged AND the per-doc summary deltas.

New file: `scripts/cosmos-back-sync.write-path.test.mjs` (architect's split rationale: different fragility profile — couples to the `@azure/cosmos` SDK call shape and turns over on SDK upgrades; the existing `cosmos-back-sync.test.mjs` is pure-function coverage and turns over only on helper changes). `syncDocument` is now exported with an `// Exported for tests; not part of CLI contract.` comment.

Shared `createFakeDestinationContainer` builder centralizes SDK call-shape knowledge in one place (`.item().read()/.replace()`, `.items.create()`) so a future SDK migration touches one builder rather than N tests.

13 tests, each pinning the entire summary delta via `assert.deepEqual` (not single-counter, which can mask cross-counter regressions):

1. **CREATE happy**: 404 from read → `items.create` called with source doc
2. **UPDATE happy with distinct etags**: `src-etag ≠ dst-etag` fixtures pin that `IfMatch.condition` is the **destination** `_etag`, NOT the source — direct regression guard against the source-etag swap
3. **skip-old-fresher**: destination newer → no write, conflict logged
4. **412 numeric from replace**: classified `concurrent-write`, `docsUpdated` NOT bumped, `conflicts` bumped
5. **409 numeric from replace**: classified `concurrent-write`
6. **`'PreconditionFailed'` string alias from replace**: errorCode normalizes to numeric 412 (composition test between `getErrorCode` alias map and the `errorCode === 412 || errorCode === 409` classifier branch — without this, a regression that removes the alias would silently reclassify a real 412 as `'unknown'`)
7. **500 from replace**: classified `unknown`
8. **`items.create` throws 409**: classified `concurrent-write` (covers the create-path branch of the same inner try/catch)
9. **dry-run UPDATE**: counter bumps, no `replace` call
10. **dry-run CREATE**: counter bumps, no `items.create` call
11. **pre-cutover filter**: `_ts < cutover` → no read/write/summary change
12. **`readDestinationDocument` 500 (not 404)**: `inspect`/`failed`/`unknown`, no write attempted
13. **malformed source (missing id)**: outer try/catch logs `'malformed-source'`, no write attempted (also covers iter-1's outer try/catch addition)

Helper-level tests in `cosmos-back-sync.test.mjs` are unchanged — architect's "do not thin direct helper-level tests when adding integration-level tests; they catch different regressions" applies.

Commit: `412d971`.

File: docs/migration-westus2.md [hunk 5]
```diff
@@ -960,6 +960,24 @@ loss of:
...
+  migration's containers. Affected containers: `blobs`,
+  `history`, `rule-sets`. The `users` container has no delete
+  path in production code and is unaffected. Resurrected
+  deletes produce zero rows in the back-sync conflicts file --
+  the script's audit trail is incomplete by design for this
+  failure mode. After the back-sync run, perform a per-container
+  diff: list document IDs that exist on OLD but not on NEW
+  (after the back-sync writes complete), filter to IDs the user
+  could have deleted post-cutover, and either re-delete them or
+  surface them for manual review.
```
[the remark this answers is not in the document]
Comment (geevensingh) [reply docs/migration-westus2.md:991]: Good catch — fixed. The Phase 3 rollback invocation now includes all three required flags:

```
node scripts/cosmos-back-sync.mjs \
  --src cosmos-jotjson-prod \
  --src-rg rg-jotjson-prod \
  --dst cosmos-jotjson-dev \
  --dst-rg rg-jotjson-dev \
  --cutover-instant-unix-seconds <ts> \
  --accept-delete-loss
```

Added a one-line note that `--accept-delete-loss` is the operator's explicit acknowledgment of the "delete-loss budget" item documented in the same Phase 3 section (standard Cosmos change feed does not replay deletes).

Verified there's only one operator invocation in the runbook; the only other appearances of `cosmos-back-sync.mjs` are the script's own USAGE text (already correct) and code/test references.

Commit: `0128f9b`.

File: scripts/cosmos-back-sync.mjs [hunk 6]
```diff
@@ -0,0 +1,894 @@
...
+//        on oldDoc._etag. 412 -> log to conflicts file.
+//     4. If 200 and oldDoc._ts >= newDoc._ts: skip + log 'old-fresher'.
+//
+// Conflicts file:
+//   cosmos-back-sync-conflicts-<ISO8601>.jsonl in CWD.
+//   --dry-run uses cosmos-back-sync-conflicts-dryrun-<ISO8601>.jsonl.
+//   One JSON object per line. Fields:
+//     {container, id, partitionKey, reason, oldTs, newTs,
+//      oldEtag, newEtag, attemptedAt}
+//   reason: 'old-fresher' | 'concurrent-write' | 'unknown'.
```
[the remark this answers is not in the document]
Comment (geevensingh) [reply scripts/cosmos-back-sync.mjs]: Good catch — fixed. `'malformed-source'` was a documentation gap; the script emits it correctly from `syncDocument`'s outer try/catch (added in iter-1 to keep one bad source doc from crashing the entire rollback) but the contract documentation enumerated only three reasons.

Updated both authoritative listings to enumerate all four reasons with one-line operator-reconciliation hints:

1. **`scripts/cosmos-back-sync.mjs`** header (the contract that ships with the script):
```
reason:
  'old-fresher'       - oldDoc._ts >= newDoc._ts; left in place.
  'concurrent-write'  - 412/409 from replace/items.create.
  'malformed-source'  - source doc failed id / partition-key /
                        _ts validation; logged with id
                        '<unknown>' when the id itself is bad.
                        Operators reconciling rollbacks must
                        investigate these rows manually because
                        the script could not classify them.
  'unknown'           - other errors (e.g. 5xx from read or
                        write); manual investigation required.
```

2. **`docs/migration-westus2.md`** PR-D conflicts-file-format section: parallel update so the runbook matches the in-script contract.

Verified those are the only two reason enumerations in the repo; tests reference individual reason values but are not contract docs.

Commit: `0128f9b`.

File: scripts/cosmos-back-sync.mjs [hunk 7]
```diff
@@ -0,0 +1,904 @@
...
+  containerName,
+  destinationContainer,
+  partitionKeyPaths,
+  cutoverInstantUnixSeconds,
+  dryRun,
+  conflictsFilePath,
+  summary,
+}) {
+  if (!Number.isFinite(sourceDocument?._ts) || sourceDocument._ts < cutoverInstantUnixSeconds) {
+    return;
```
[the remark this answers is not in the document]
Comment (geevensingh) [reply scripts/cosmos-back-sync.mjs]: Fixed in 719cff4.

You're right — and thanks for catching the drift one iteration later. Iter-3 (commit 0128f9b) shipped the contract enumeration mentioning `_ts` validation in both the script header and the runbook *before* the code matched. This iter-4 commit makes that promise true.

**Shape of the fix** (per a 3-agent rubber-duck panel on the change — skeptic + advocate + architect; 13 distinct findings, 11 ADOPT, 3 SET ASIDE with architectural reasons):

- The line-489 combined guard `!Number.isFinite(_ts) || _ts < cutover` was conflating two distinct concerns (silent pre-cutover scope filter vs. malformed-source classification). Split into:
  - Pre-cutover filter (silent skip) requires `Number.isFinite(_ts) && _ts < cutover` — only finite-numeric pre-cutover docs are silently dropped.
  - A new explicit throw inside the outer try/catch, placed AFTER the id check and BEFORE `extractPartitionKeyValue`. The ordering matters: missing-`_ts` on a multi-defect doc shouldn't be masked by an unrelated partition-key throw, so operators see the most-actionable failure first in the conflicts file.
- `decideAction`'s own `Number.isFinite(newDoc._ts)` check is **kept** as defense-in-depth at the unit boundary. With the upstream fix, that check is unreachable through `syncDocument` but it's still exercised by direct unit tests in `cosmos-back-sync.test.mjs`. Removing it would mean `decideAction(oldDoc, { _ts: NaN })` returns `'skip-old-fresher'` (because `NaN < anything === false`) — silently wrong, worse than throwing. Added a JSDoc note on `decideAction` explaining the duplication.

**Tests added** (9 new in `cosmos-back-sync.write-path.test.mjs`):

- Loop-parameterized 6 malformed-`_ts` cases: `missing`, `'1234'`, `NaN`, `Infinity`, `-Infinity`, `null`. Each pins: no read/replace/create, full summary delta via `deepEqual` (docsProcessed=0, docsMalformed=1, conflicts=1), conflict reason='malformed-source', conflict id = actual source id (not `<unknown>` — id check fires first), and `errorCode` matches `/missing a numeric _ts/` (pinned so runbook `jq` examples don't silently rot).
- **pre-cutover × non-numeric `_ts`**: regression-prevention test that fails if someone reverts the `&&` to `||`.
- **`_ts === cutoverInstantUnixSeconds` exact boundary**: pins strict-less-than semantics.
- **id missing AND `_ts` NaN**: pins inside-try ordering — conflict row id is `<unknown>` because id check fires first.

The iter-3 runbook hint at line 318 ("source doc failed id / partition-key / `_ts` validation; operator must investigate manually") becomes accurate as of this commit; no doc edits needed.

DoD: `npm run lint:all`, `npm run test:scripts` (468/468), `npm --prefix api test` (466/466) all green.
